File: backend/wsp/src/main/java/wsp/entity/Recipe.java
package wsp.entity;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "recipes")
@Getter
@Setter
public class Recipe {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(length = 2000)
    private String description;

    @Column(nullable = false, length = 5000)
    private String instructions;

    @ManyToOne(optional = false)
    @JoinColumn(name = "group_id", nullable = false)
    private UserGroup group;

    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RecipeIngredient> ingredients = new ArrayList<>();

    public void addIngredient(RecipeIngredient ingredient) {
        ingredients.add(ingredient);
        ingredient.setRecipe(this);
    }

    public void clearIngredients() {
        ingredients.forEach(ingredient -> ingredient.setRecipe(null));
        ingredients.clear();
    }
}

File: client/wsp/lib/features/recipes/models/recipe.dart
import 'package:wsp/features/recipes/models/recipe_ingredient.dart';

class Recipe {
  const Recipe({
    required this.id,
    required this.groupId,
    required this.title,
    this.description,
    required this.instructions,
    required this.ingredients,
  });

  final int id;
  final int groupId;
  final String title;
  final String? description;
  final String instructions;
  final List<RecipeIngredient> ingredients;

  factory Recipe.fromJson(Map<String, dynamic> json) {
    final ingredientsJson = json['ingredients'] as List<dynamic>? ?? [];

    return Recipe(
      id: json['id'] as int,
      groupId: json['groupId'] as int,
      title: json['title'] as String,
      description: json['description'] as String?,
      instructions: json['instructions'] as String,
      ingredients: ingredientsJson
          .map(
            (item) => RecipeIngredient.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
    );
  }
}
